plugins/renepay/mcf.c

```diff
@@ -0,0 +1,1507 @@
+ * On the other hand, we are now also considering Pickhard's
+ * [1] model to improve payment reliability,
+ * hence our optimization moves to a 2D space: either we like to maximize the
+ * probability of success of a payment or minimize the routing fees, or
+ * alternatively we construct a function of the two that gives a good compromise.
+ *
+ * Therefore from now own, the definition of optimal is a matter of choice.
+ * To simplify the API of this module, we think the best way to state the
+ * problem is:
+ *
+ * 	Find a routing solution that pays the least of fees while keeping
+ * 	the probability of success above a certain value `min_probability`.
+ *
+ *
+ * # Fee Cost
+ *
+ * Routing fees is non-linear function of the payment flow x, that's true even
+ * without the base fee:
+ *
+ * 	fee_msat = base_msat + floor(millionths*x_msat / 10^6)
+ *
+ * We approximate this fee into a linear function by computing a slope `c_fee` such
+ * that:
+ *
+ * 	fee_microsat = c_fee * x_sat
+ *
+ * Function `linear_fee_cost` computes `c_fee` based on the base and
+ * proportional fees of a channel.
+ * The final product if microsat because if only
+ * the proportional fee was considered we can have c_fee = millionths.
+ * Moving to costs based in msats means we have to either truncate payments
+ * below 1ksats or estimate as 0 cost for channels with less than 1000ppm.
+ *
+ * TODO(eduardo): shall we build a linear cost function in msats?
+ *
+ * # Probability cost
+ *
+ * The probability of success P of the payment is the product of the prob. of
+ * success of forwarding parts of the payment over all routing channels. This
+ * problem is separable if we log it, and since we would like to increase P,
+ * then we can seek to minimize -log(P), and that's our prob. cost function [1].
+ *
+ * 	- log P = sum_{i} - log P_i
+ *
+ * The probability of success `P_i` of sending some flow `x` on a channel with
+ * liquidity l in the range a<=l<b is
+ *
+ * 	P_{a,b}(x) = (b-x)/(b-a); for x > a
+ * 		   = 1.         ; for x <= a
+ *
+ * Notice that unlike the similar formula in [1], the one we propose does not
+ * contain the quantization shot noise for counting states. The formula remains
+ * valid independently of the liquidity units (sats or msats).
+ *
+ * The cost associated to probability P is then -k log P, where k is some
+ * constant. For k=1 we get the following table:
+ *
+ * 	prob | cost
+ * 	-----------
+ * 	0.01 | 4.6
+ * 	0.02 | 3.9
+ * 	0.05 | 3.0
+ * 	0.10 | 2.3
+ * 	0.20 | 1.6
+ * 	0.50 | 0.69
+ * 	0.80 | 0.22
+ * 	0.90 | 0.10
+ * 	0.95 | 0.05
+ * 	0.98 | 0.02
+ * 	0.99 | 0.01
+ *
+ * Clearly -log P(x) is non-linear; we try to linearize it piecewise:
+ * split the channel into 4 arcs representing 4 liquidity regions:
+ *
+ * 	arc_0 -> [0, a)
+ * 	arc_1 -> [a, a+(b-a)*f1)
+ * 	arc_2 -> [a+(b-a)*f1, a+(b-a)*f2)
+ * 	arc_3 -> [a+(b-a)*f2, a+(b-a)*f3)
+ *
+ * where f1 = 0.5, f2 = 0.8, f3 = 0.95;
```

Comment:
I love the non uniform split of the capacity! However I believe 0.5 is a pretty high probability. I fear this will trigger and saturate too large payments. I'd love to have a discussion on the choices. also one does not need to take a uniform distribution for the probability to begin with. LND uses a a bimodal distribution. However that it hard to linearize.

Reply:
These are more parameters can we have to tune with experience.

Reply:
@renepickhardt, there is another detail here that bothers me, it is the fact that we cut the tail of the distribution at 95% of the channel capacity. But need either to add more arcs above that increasing the computational and memory burden, or produce a less precise linearization, for example like extending the last bin from 80% to 99%.